### PR TITLE
Feat/cover adapter

### LINKED ISSUE #34: Cover Adapter

### PR DESCRIPTION
**Pull request by kseniya292**
Closes #34

### CODE REVIEW

File: contracts/SaveToken.sol [hunk 1]
```diff
@@ -140,12 +142,14 @@ contract SaveToken is ERC20Extended, Pausable {
             _delegatecall(StorageLib.assetAdapter(), signature_transfer);
         }
 
+        uint256 balance = super.balanceOf(msg.sender);
```
Comment (dougiebuckets) [inline contracts/SaveToken.sol]: Thoughts on moving these into `_transferTokens`, so we don't need to include them in both `transfer` and `transferFrom`? Would also prevent us from needing to pass around the additional `ratio` argument.

File: contracts/SaveToken.sol [hunk 2]
```diff
@@ -257,9 +262,32 @@ contract SaveToken is ERC20Extended, Pausable {
...
+        address sender,
+        address recipient,
+        uint256 ratio)
+        internal 
+        {
+        
+        uint256 assetBalance = StorageLib.getAssetBalance(sender);
+        uint256 insuranceBalance = StorageLib.getInsuranceBalance(sender);
+
+        uint256 assetTransferAmount = ratio.mul(assetBalance);
```
Comment (dougiebuckets) [inline contracts/SaveToken.sol]: We might want to include quick comment here that explains why we need to use the ratio.

File: contracts/adapters/CoverAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "../libraries/StorageLib.sol";
+import "../interfaces/external/IBalancerPool.sol";
+import "../interfaces/external/ICoverToken.sol";
+import "../interfaces/IInsurance.sol";
+
+contract CoverAdapter is IInsurance {
+    using SafeMath for uint256;
```
Comment (dougiebuckets) [inline contracts/adapters/CoverAdapter.sol:12]: Now that OZ supports Solidity 0.8 I think we can go ahead and update our contracts and remove SafeMath. Went ahead and added a ticket: https://github.com/save-dai/savetoken/issues/47

File: contracts/adapters/CoverAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        // approve the transfer
+        underlyingToken.approve(address(balancerPool), amount );
+
+        // solhint-disable-next-line
+        (uint256 tokensOut,) = balancerPool.swapExactAmountIn(
+            address(underlyingToken), // tokenIn
+            amount, // tokenAmountIn
+            address(covToken), // tokenOut
+            1, // minAmountOut
+            type(uint256).max // maxPrice
```
Comment (dougiebuckets) [inline contracts/adapters/CoverAdapter.sol:32]: Cool! Hadn't seen these before.

File: contracts/libraries/StorageLib.sol [hunk 5]
```diff
@@ -100,8 +100,8 @@ library StorageLib {
         assetToken_ = saveTokenStorage().insuranceToken;
     }
 
-    function uniswapFactory() internal view returns (address uniswapFactory_) {
-        uniswapFactory_ = saveTokenStorage().uniswapFactory;
+    function exchangeFactory() internal view returns (address uniswapFactory_) {
+        uniswapFactory_ = saveTokenStorage().exchangeFactory;
```
Comment (dougiebuckets) [inline contracts/libraries/StorageLib.sol]: I'll replace these remaining `uniswapFactory` references here and in `setAddresses` in my upcoming PR.

**Review by dougiebuckets — COMMENTED** [review]
Looks great @kseniya292! A couple small notes we can discuss as I wrap up tests for `withdrawForUnderlyingAsset`.

File: contracts/SaveToken.sol (hunk 1 above)
Comment (kseniya292) [reply contracts/SaveToken.sol]: @dougiebuckets good catch. Updating.

File: contracts/adapters/CoverAdapter.sol (hunk 3 above)
Comment (kseniya292) [reply contracts/adapters/CoverAdapter.sol:12]: 👍

**Pull request merged**